### PR TITLE
gh-117784: Only reference PHA functions ifndef SSL_VERIFY_POST_HANDSHAKE

### PR DESCRIPTION
**Pull request by WillChilds-Klein**
# Notes

AWS-LC [recently defined](https://github.com/aws/aws-lc/pull/1526) a new configuration macro `OPENSSL_NO_TLS_PHA` to signal lack of support for [TLSv1.3 post-handshake authentication](https://datatracker.ietf.org/doc/html/rfc8446#section-4.2.6). Other cryptography libraries [may find this useful](https://bugs.chromium.org/p/boringssl/issues/detail?id=78). We then adjust `Modules/_ssl.c` to detect this macro, guarding usages of PHA-related functions. We also update two PSK tests to generate client/server contexts with `test_ssl.py`'s conventional `testing_context()` utility.

This pull request implements Issue #117784.

* Issue: gh-117784

### CODE REVIEW

**Comment by WillChilds-Klein** [comment]
My [apologies](https://github.com/python/cpython/pull/117234#issuecomment-2027789747) on the [force-push](https://github.com/python/cpython/compare/4d1511923b9d19e29367939fadda299077fea840..4c8cee2a84ec8f2676eb85ebfa70f8d8885bfcce) in this draft. I inadvertently rebased latest from upstream main. No significant commit history was lost; I'll refrain from rebasing main again on this PR.

**Comment by WillChilds-Klein** [comment]
Hi @encukou and @gpshead, do you have any thoughts on this PR?

**Comment by encukou** [comment]
I'm not a SSL expert, but I'm concerned about relying on an `OPENSSL_*` macro that doesn't come from the OpenSSL project. Not sure if we have any precedent for that -- it looks like the other OpenSSL-only things use `#ifdef` with macros that OpenSSL defines.

Does AWS-LC still define `SSL_VERIFY_POST_HANDSHAKE`?

**Comment by WillChilds-Klein** [comment]
Good call @encukou. Neither [AWS-LC](https://github.com/search?q=repo%3Aaws%2Faws-lc+SSL_VERIFY_POST_HANDSHAKE&type=code) nor [BoringSSL ](https://github.com/search?q=repo%3Agoogle%2Fboringssl+SSL_VERIFY_POST_HANDSHAKE&type=code) define `SSL_VERIFY_POST_HANDSHAKE`, so lack of that macro can be used to detect lack of PHA support. I've updated AWS-LC's CI and this PR accordingly.

File: Lib/test/test_ssl.py [hunk 1]
```diff
@@ -4364,14 +4364,14 @@ def test_session_handling(self):
     def test_psk(self):
         psk = bytes.fromhex('deadbeef')
 
-        client_context = ssl.SSLContext(ssl.PROTOCOL_TLS_CLIENT)
+        client_context, server_context, _ = testing_context()
```
Comment (encukou) [inline Lib/test/test_ssl.py]: Is this change needed?
Not being an OpenSSL expert, I'd rather avoid having to review it :)

File: Modules/_ssl.c [hunk 2]
```diff
@@ -187,6 +187,11 @@ extern const SSL_METHOD *TLSv1_2_method(void);
 #endif
 
 
+#if !defined(SSL_VERIFY_POST_HANDSHAKE) || !defined(TLS1_3_VERSION) || defined(OPENSSL_NO_TLS1_3)
+  #define PY_SSL_NO_POST_HS_AUTH
```
Comment (encukou) [inline Modules/_ssl.c]: I'd prefer a positive name: it is always used with `!` later and double negatives are harder to read.

Perhaps `PySSL_HAVE_POST_HS_AUTH`

File: Modules/_ssl.c [hunk 3]
```diff
@@ -3576,7 +3582,7 @@ set_maximum_version(PySSLContext *self, PyObject *arg, void *c)
     return set_min_max_proto_version(self, arg, 1);
 }
 
-#ifdef TLS1_3_VERSION
+#if defined(TLS1_3_VERSION) && !defined(OPENSSL_NO_TLS1_3)
```
Comment (encukou) [inline Modules/_ssl.c:3585]: Are tickets used only for post-handshake authentication? From the docs it looks like a separate TLSv1.3 feature.

File: Lib/test/test_ssl.py (hunk 1 above)
Comment (WillChilds-Klein) [reply Lib/test/test_ssl.py]: It's required for this test to pass when built against AWS-LC, but I'll remove it to conform to your guidance [here](https://discuss.python.org/t/support-building-ssl-and-hashlib-modules-against-aws-lc/44505/4):

> IMO we could get to the point where you can get CPython to compile with AWS-LC without special-casing it (i.e. no Py_OPENSSL_IS_AWSLC), but relevant test cases still fail.
> You’ll still need a patch, but a more focused one.

File: Modules/_ssl.c (hunk 3 above)
Comment (WillChilds-Klein) [reply Modules/_ssl.c:3585]: No, tickets are used for session resumption and are unrelated to post-handshake authentication. I believe `get_num_tickets` pertains to [a post-handshake message added in TLSv1.3](https://www.rfc-editor.org/rfc/rfc8446.html#section-4.6.1). `SSL_CTX_get_num_tickets`'s [docs](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_get_num_tickets.html) indicate as much.

This change just updates that guard to honor [OpenSSL's `OPENSSL_NO_TLS1_3`](https://github.com/search?q=repo%3Aopenssl%2Fopenssl+OPENSSL_NO_TLS1_3&type=code) as [other parts](https://github.com/python/cpython/blame/92893fd8dc803ed7cdde55d29d25f84ccb5e3ef0/Modules/_ssl.c#L253) of `_ssl.c` already do.

**Review by encukou — APPROVED** [review]

**Comment by WillChilds-Klein** [comment]
Thank you @encukou! Would it be possible to backport this to 3.13? I see that 3.12 is [currently in "bugfix"](https://devguide.python.org/versions/), so I assume this patch isn't applicable to <= 3.12.
